packages/create-toolpad-app/src/templates/auth/auth.ts

```diff
@@ -0,0 +1,74 @@
+import { ProvidersTemplate } from '../../types';
+
+const auth: ProvidersTemplate = (providers) => {
+  const hasCredentialsProvider = providers.includes('Credentials');
+  const oAuthProviders = providers.filter((provider) => provider !== 'Credentials');
+
+  return `import NextAuth from 'next-auth';
+  ${providers
+    .map((provider) => `import ${provider} from 'next-auth/providers/${provider?.toLowerCase()}';`)
+    .join('\n')}
+  import type { Provider } from 'next-auth/providers';
+
+  const providers: Provider[] = [
+    ${oAuthProviders
+      .map(
+        (provider) => `${provider}({
```

Comment:
You could branch here on whether it's the credentials provider to avoid having to traverse the provider array twice in
```tsx
const hasCredentialsProvider = providers.includes('Credentials');
const oAuthProviders = providers.filter((provider) => provider !== 'Credentials');
```